homie-controller/src/types.rs

```diff
@@ -245,6 +278,45 @@ pub struct Device {
 
     /// The nodes of the device, keyed by their IDs.
     pub nodes: HashMap<String, Node>,
+
+    /// The Homie extensions implemented by the device.
+    pub extensions: Vec<Extension>,
+
+    /// The IP address of the device on the local network.
+    pub local_ip: Option<String>,
```

Comment:
How big is the homie spec? Is it worth having a `HashMap<String, String>` of overflow properties?

I suppose that might be hard to do in a backwards-compatible way, because as soon as you add first-class support for something, it stops existing in the overflow properties.

Reply:
... also: how big is this struct in bytes? How big and sparse does it need to get before it would be more efficient to store everything in a hashmap and expose getters for each of the properties? ("we don't expect to ever handle enough devices to care" is a valid answer to these questions)

Reply:
I would prefer them to be typed, though some way to separate extensions (like the ones added in this PR) from the core spec might be good. I'm not sure how best to do that though, or whether it's actually worthwhile.

Reply:
Might be interesting to separate the extensions out with traits. Not needed for the current feature set though, so we have some time to ponder how to do it.